### PR TITLE
feat(170): grading and checking tests

### PR DESCRIPTION
**Pull request by mrstrval**

### CODE REVIEW

File: server/db-middleware/check-grade-api.js [hunk 1]
```diff
@@ -32,28 +39,142 @@ const isGraidingPossibleIndividual = function (assignment) {
...
+    .then((check) => {
+      requestAmount = !check;
     })
     .then(() => TestSubmission
       .countDocuments({ assignmentId: assignment._id, status: REQUEST_STATUS_CHECKED }))
     .then((submissionAmount) => {
-      if (stdAmount === submissionAmount ||
-        (assignment.deadline !== null && assignment.deadline < new Date().getTime())
+      if ((stdAmount === submissionAmount && requestAmount) ||
+        (assignment.deadline !== undefined && assignment.deadline < new Date().getTime())
```
Comment (whybazinga) [inline server/db-middleware/check-grade-api.js]: !assignment.deadline

File: server/db-middleware/check-grade-api.js [hunk 2]
```diff
@@ -32,28 +39,142 @@ const isGraidingPossibleIndividual = function (assignment) {
...
 };
 
-apiModule.isGraidingPossible = function (assignmentId) {
+const updateQuestion = function (questId) {
+  return Question
+    .findById(questId)
+    .select('peopleTested peopleAnswered difficulty')
+    .then((info) => {
+      let diff;
+      if (+info.peopleTested !== 0) {
```
Comment (whybazinga) [inline server/db-middleware/check-grade-api.js]: No need in unary operator, it is specified in schema as number, so no casting required

File: server/db-middleware/check-grade-api.js [hunk 3]
```diff
@@ -32,28 +39,142 @@ const isGraidingPossibleIndividual = function (assignment) {
...
+        diff = info.peopleTested - info.peopleAnswered;
+        diff /= +info.peopleTested;
```
Comment (whybazinga) [inline server/db-middleware/check-grade-api.js]: same issue with unary operator

File: server/db-middleware/check-grade-api.js [hunk 4]
```diff
@@ -32,28 +39,142 @@ const isGraidingPossibleIndividual = function (assignment) {
...
+        diff *= 4;
+        diff = Math.round(0.5 + diff);
+        if (diff > 4) diff = 4;
+        else if (diff < 1) diff = 1;
+      }
+      if ((+info.peopleTested > minToChangeMark) &&
```
Comment (whybazinga) [inline server/db-middleware/check-grade-api.js]: same

**Review by whybazinga — CHANGES_REQUESTED** [review]

File: server/db-middleware/submission-api.js [hunk 5]
```diff
@@ -819,21 +848,23 @@ apiModule.sendCheckingResults = function (checkingId, res) {
...
-        elReassign.result = res;
-        return elReassign;
-      }
-      return el;
-    })))
+    .then((submission) => {
+      updateQuestion(question, res);
+      return Promise.all(submission.answers.map((el) => {
+        if (el.questionId == question) {
+          const elReassign = {};
```
Comment (annakisel) [inline server/db-middleware/submission-api.js]: why not const elReassign = { checking: ..., .........}

Comment (annakisel) [reply server/db-middleware/submission-api.js]: the you can write your whole promise like el.questionId == question ? { checking: ......, } : el;

File: server/routes/teacher/test.js [hunk 6]
```diff
@@ -106,8 +106,8 @@ route.get('/questions-check/:teacherId', (req, res) => {
 route.get('/check-res/:checkid/:result', (req, res) => {
```
Comment (annakisel) [inline server/routes/teacher/test.js]: post

**Review by whybazinga — APPROVED** [review]

**Review by annakisel — APPROVED** [review]

**Pull request merged**